### ISSUE TITLE
demo: `cockroach demo` randomizes listen address on Ubuntu when `--insecure` is used

### ISSUE THREAD

**Issue by WadeWaldron** [cockroachdb/cockroach#68437]
**Describe the problem**

I created a VM for Ubuntu 20.04 LTS.

On the VM I ran the following command:

```
cockroach demo --insecure --sql-port 26257
```

When I run this command, it listens on a random address.

Now, if I compare this to running on MacOS Big Sur, on Mac it always defaults to listening on 127.0.0.1 (localhost).

It's worth noting that if I remove the `--insecure` option, then it doesn't randomize the listen address anymore.

**To Reproduce**

Run the above command on Ubuntu 20.04 LTS.

**Expected behavior**

Both Mac and Linux should have a consistent behaviour. I should be able to provide instructions for one platform and expect they will work the same on another.

I would say that ideally, it should default to localhost for demo mode.

There should also be an option such as `--listen-addr` to allow me to configure the specific address to that I can guarantee it will be correct.

**Environment:**

- CockroachDB version: 21.1.6 (Linux), 21.1.2 (Mac) 
- Server OS: Ubuntu 20.04, Mac OS 11.5 (Big Sur)

**Additional context**

The impact of this is that when I attempt to build instructions and/or scripts that use Demo mode to show off features of Cockroach DB, I need to take into account what platform a user might be working with. On Linux the instructions get a lot more complicated because they can't rely on a consistent address.

Jira issue: CRDB-9035

Epic CRDB-22814

**Comment by knz** [comment]
@WadeWaldron can you illustrate what you mean with "different address". Screenshot? evidence?

**Comment by WadeWaldron** [comment]
On Mac:

```
> cockroach demo --insecure --sql-port 26257
...
# Connection parameters:
#   (webui)    http://127.0.0.1:8080
#   (sql)      postgres://root@127.0.0.1:26257?sslmode=disable
#   (sql/unix) postgres://root:unused@?host=%2Fvar%2Ffolders%2Fs1%2F0099t4sj6r78qljb2k5knrhh0000gq%2FT%2Fdemo149454343&port=26257
...
```

On Ubuntu:

```
...
> cockroach demo --insecure --sql-port 26257
#   - Connection parameters:
#     (webui)    http://127.32.248.149:8080 <-- The IP changes every time I execute.
#     (sql)      postgres://root@127.32.248.149:26257?sslmode=disable
#     (sql/unix) postgres://root:unused@?host=%2Ftmp%2Fdemo125056151&port=26257
...
```

```
> cockroach demo --sql-port 26257
...
#   - Connection parameters:
#     (webui)    http://127.0.0.1:8080/demologin?password=demo100976&username=demo
#     (sql)      postgres://demo:demo100976@127.0.0.1:26257?sslmode=require
#     (sql/unix) postgres://demo:demo100976@?host=%2Ftmp%2Fdemo575595135&port=26257
...
```

**Comment by knz** [comment]
ok let's look into it

meanwhile:
1) the `--insecure` flag is deprecated.
2) we *really* do not want to advertise its usage in user-facing docs. 

Ideally you'd be able to build whatever material you're currently working on without it. You mentioned inconvenience at the changing password. Have you considered asking the demo shell to create another username with a stable password instead?

**Comment by WadeWaldron** [comment]
Wasn't aware that `--insecure` was deprecated. Thanks for the heads up.

So what is the replacement for `--insecure` when using `cockroach demo`?

Is there command line params that allow you to set a user/password? Or am I going to have to use the randomized password to login and then create a user and password?

**Comment by knz** [comment]
The fact that the docs don't specify it is deprecated is a political game around a catch-22:

- the idea of an insecure mode is 100% misaligned with our business, so we don't want it
- there are certain use cases which due to convenience started mistakenly relying on the insecure mode
- because these use cases are now satisfied, we do not experience product pressure to address these use cases in different ways
- the docs team is opposed to announcing the deprecation because there is no alternate way available for these use cases

If it was just me, we would remove the feature entirely, see what breaks / what is inconvenienced, then build the right tooling again for those use cases. Using a currently broken situation as leverage to block the removal of something misaligned with our business is bad optics all arounds.

cc @piyush-singh

**Comment by piyush-singh** [comment]
> So what is the replacement for --insecure when using cockroach demo?

When you use `cockroach demo` without the `--insecure` flag, note that credentials are generated for you:

```
12:46 $ cockroach demo
#
# Welcome to the CockroachDB demo database!
#
# You are connected to a temporary, in-memory CockroachDB cluster of 1 node.
#
# This demo session will attempt to enable enterprise features
# by acquiring a temporary license from Cockroach Labs in the background.
# To disable this behavior, set the environment variable
# COCKROACH_SKIP_ENABLING_DIAGNOSTIC_REPORTING=true.
#
# Beginning initialization of the movr dataset, please wait...
#
# The cluster has been preloaded with the "movr" dataset
# (MovR is a fictional vehicle sharing company).
#
# Reminder: your changes to data stored in the demo session will not be saved!
#
# If you wish to access this demo cluster using another tool, you will need
# the following details:
#
#   - Connection parameters:
#     (webui)    http://127.0.0.1:8080/demologin?password=demo50721&username=demo
#     (sql)      postgres://demo:demo50721@127.0.0.1:26257?sslmode=require
#     (sql/unix) postgres://demo:demo50721@?host=%2Fvar%2Ffolders%2F32%2F2yps592x5c17r8byy9lfd21h0000gn%2FT%2Fdemo083536332&port=26257
#   
#   - Username: "demo", password: "demo50721"
#   - Directory with certificate files (for certain SQL drivers/tools): /var/folders/32/2yps592x5c17r8byy9lfd21h0000gn/T/demo083536332
#
# Server version: CockroachDB CCL v21.1.6 (x86_64-apple-darwin19, built 2021/07/20 15:33:43, go1.15.11) (same version as client)
# Cluster ID: e3663542-5187-4c72-94eb-1f0f023a439b
# Organization: Cockroach Demo
#
# Enter \? for a brief introduction.
#
demo@127.0.0.1:26257/movr>
```

Note both:
- Auto-generated username and pw: `- Username: "demo", password: "demo50721"`
- Link to the DB Console that automatically logs the user in using the temp credentials: `http://127.0.0.1:8080/demologin?password=demo50721&username=demo`

Beyond this, are there additional things you are looking for @WadeWaldron? If there are other use cases that you'd like to see supported in the secure version of demo, let us know and we can file issues for them. Our goal is to make the secure UX as smooth and easy to use as the `--insecure` UX.

**Comment by WadeWaldron** [comment]
For my purposes, I need a way to provide a consistent, predefined way for an application to connect to a demo cluster.

In other words, I want a user to be able to run the demo cluster, and then execute an application and connect to it, without them having to update config files etc.

I currently can't do that because:
- `--insecure` randomizes the ip address.
- Removing `--insecure` randomizes the password.

Basically, I'm looking to eliminate the randomization.

One way to do this would be a command like:

```
cockroach demo --host 127.0.0.1 --sql-port 36257 --default-sql-user movr --default-sql-password password
```

This would allow me to guarantee that there is always a consistent host/port/user/password combination. I could then code that into my config files so that I can ensure that the application will successfully connect to the database without any issues.

Alternatively, having demo mode always provide a consistent default host/port/user/password combination without any additional configs would also work.

**Comment by knz** [comment]
Have you considered `start-single-node` with the `--listen-url` flag?

Alternatively we could make `--listen-url` work with `demo`

Yet another alternative is to run `demo -e 'CREATE USER ...' -e 'SELECT pg_sleep(1000000)'` or some larger value.

**Comment by WadeWaldron** [comment]
`start-single-node` doesn't work. I need the enterprise features enabled.

I have a workaround to initialize a specific user/password. But it would be nice if I didn't need the workaround.

**Comment by knz** [comment]
I think we're letting the tail wag the dog here.

At a very high level, it looks to me that you need a *regular cluster* (i.e. not a `demo` cluster) but also need to issue it a temporary enterprise license. Why not do that instead? Get an education-only CRL license and use that for all your education purposes?

**Comment by WadeWaldron** [comment]
What I need is a self-serve way for users to access Cockroach Enterprise features. It doesn't matter to me how they get it. But it has to be self-serve. Ideally this would actually be done through Cockroach Cloud, but currently that's not possible (as far as I know).

A temporary education license isn't self-serve, unless we make the license permanent, which is not a good idea.

I'm open to other ideas, but at the moment, Cockroach Demo is all I have.

**Comment by WadeWaldron** [comment]
Having said all of that. Whether I had access to those things or not, I still think that demo mode could benefit from having a way to start it up with a consistent host/port/user/password combination.

**Comment by knz** [comment]
> What I need is a self-serve (and no credit card required) way for users to access Cockroach Enterprise features.

This does not sound well-aligned with Cockroach Labs' overall business. How do we ensure that this facility, if it ever existed, wouldn't undercut our CC  offering altogether?

**Comment by knz** [comment]
I also recall Jesse had success instantiating Katakoda instances for a similar purpose, without the risk to expose arbitrary licenses to users. What was the problem with reusing that for other education projects?

**Comment by WadeWaldron** [comment]
I think we're off track.

The original issue was that demo mode behaves differently on Linux than it does on Mac.

Regardless of how or why I want to use it, I think we should have an option to run Cockroach Demo in a manner that is consistent across platforms and across executions. I.E. There should be a way to ensure that Cockroach Demo always executes with the same host/port/user/password.

As to why we should need that? If someone wants to run Cockroach locally, and they want to experiment with Enterprise features, then demo seems like the way to do that. If they are doing these experiments, then having to constantly copy/paste new urls/passwords etc. because they restarted the DB (maybe because they messed something up) is kind of annoying. Would be nice if we could avoid that.

**Comment by knz** [comment]
The purpose and design of `cockroach demo` is *specifically* to provide an evaluation environment *for interactive use only* - i.e. it was a strong requirement to limit the ability of `cockroach demo` to start up a cluster in a way that would behave and serve just as well as "the real thing". So as to avoid users depending on it for development and evaluation. We have evaluation licenses for that latter purpose.

The randomized credentials are *also* a requirement, so as to avoid the risk that a user idly running `cockroach demo` on their computer mistakenly give access to malware to their files, by misusing the running crdb process (both on macOS and linux, there is no protection against a network process being connected to by another process running on the same machine).

Your use case is violating the two design requirements:

- you wish to offer "the real deal" and thus are running against the 1st design objective, and likely requiring a technical product that misaligned with the licensing business
- your request for stability is raising a couple of security questions, how to limit the risk to open vulnerabilities on the computers where this facility will run

I don't think we should treat this situation casually, and I'd definitely like to see someone from product chime in on this, as well as possibly @bdarnell

**Comment by piyush-singh** [comment]
>So as to avoid users depending on it for development and evaluation. We have evaluation licenses for that latter purpose.

I believe (please correct me if I'm wrong) Wade is trying to create courses for CockroachU. `cockroach demo`'s temporary license is handy for education/training purposes as it doesn't require the user to do anything to set it up and it expires after 1 hour. `cockroach demo` also has the nice property of storing everything in memory which makes cleanup after a lesson a simple `ctrl+c`.

We can issue temporary "evaluation" licenses for the education team, but those expire after one month and lock enterprise features afterwards. That wouldn't work for Wade's use case as he'd have to keep updating instructions to use new licenses on CockroachU. We also have full enterprise licenses, but those do not disable enterprise features after they expire (and still only would be able to be applied to new clusters before they expire).

>you wish to offer "the real deal" and thus are running against the 1st design objective, and likely requiring a technical product that misaligned with the licensing business

I do think `cockroach demo` is a useful way to kick the tires of CockroachDB's enterprise features and can serve as a useful tool when it comes to training. I don't think we need this to be a perfect simulacrum of real `cockroach`. Wade, please correct me if I'm mistaken here. I don't know enough about the Katakoda experiment Jesse did to know whether that is a suitable replacement.

> The randomized credentials are also a requirement, so as to avoid the risk that a user idly running cockroach demo on their computer mistakenly give access to malware to their files, by misusing the running crdb process (both on macOS and linux, there is no protection against a network process being connected to by another process running on the same machine).

I think this is the crux of this issue. If we provide a way to configure username/pw and our instructions on CockroachU pass a consistent set of credentials, it seems that it would invite abuse. Given that `cockroach demo` is only intended for short sessions, we'd need to weigh this risk against the benefit of having a consistent set of credentials that can be used in examples during training. I don't know enough about the magnitude or probability of the security risks here, perhaps @bdarnell has an opinion.

**Comment by WadeWaldron** [comment]
@piyush-singh, you've basically hit the nail on the head.

I feel like the fact that `cockroach demo` only runs in memory mitigates most of the risk. Why would anyone run that in a production environment? It's definitely not the "real deal". And there isn't really any access to files. So the risk in my mind is that someone could get access to a non-production, in-memory database. Maybe I'm missing some key issues here, but that seems like a pretty minor risk. Plenty of production software offers "default passwords" that they expect you to change if you are running in production.

The Katacoda experiment may be a way forward. We're investigating, and it (or other similar products) are definitely on the roadmap. However, I don't have access to them today. Today, I have access to `cockroach demo`.

I'm trying to avoid having to summarize weeks/months of discussions and investigations. There's a lot going on some of it may eventually make this discussion irrelevant.

In the meantime, we have the original issue which is that `cockroach demo` doesn't behave the same on Mac and Linux.

**Comment by knz** [comment]
I think the best outcome is to have `cockroach demo` support `--listen-url` like regular servers, and write a valid connection string (including random port, username and password) in the file specified. Then other apps can use that reliably.

**Comment by WadeWaldron** [comment]
I would say the best outcome is that the Linux version of `cockroach demo` should behave the same as the Mac version of `cockroach demo`. The `--listen-url` is a nice to have feature on top of that.

**Comment by knz** [comment]
You were also asking for a way to stabilize the password - I was suggesting to push against that and lean on `--listen-url` instead.

The fact the address (and port) is variable would also become moot with `--listen-url`.

**Comment by WadeWaldron** [comment]
I actually wasn't. The original issue is that Mac/Linux behave differently.

Then we got into conversations about how I would ideally like things to work based on my use case. But that's unrelated to the original bug.

**Comment by knz** [comment]
As to the reason why this randomization is not active when using TLS (in secure mode):

the origin of the randomization facility is because multiple tests running concurrently would mistakenly connect to each others' server. When running securely, each test has different TLS certs and thus the TLS authentication would fail regardless of address. When running insecurely, there is no authentication and thus no way for two test to recognize they are not meant to talk to each other. The randomization turned out to be a reliable way to block that problem.

**Comment by knz** [comment]
> It likely won't help me.

Can you tell us a bit about the automation you've build around?

**Comment by WadeWaldron** [comment]
Due to the limitations imposed by `cockroach demo`, I have worked around them using the following:

1) Run `cockroach demo` in secure mode (i.e. without `--insecure`). Particularly since it was mentioned that this feature might be deprecated. Because I am not using `--insecure` the actual URL seems to default to localhost.
2) Have users execute a script using the (sql) endpoint listed when running `cockroach demo`.
3) The script will automatically create a specific user with a specific password. This user/password can then be used by applications that need to connect to the cockroach instance. This bypasses the randomized password issue because I can use a specific user/password combination.

With these workarounds in place, there is a small inconvenience of having to run that script to create the user and having to supply the randomized username/password. However, once that is done everything works the way I want it to. And the script had to be run anyway to create the databases. So the only real inconvenience is having to supply that url rather than being able to default it. The advantage is that this is a more realistic usage (i.e. creating a user for this purpose is probably what you would do in a production environment).

With the `--listening-url-file` it's unclear to me at the moment how I would get the data from that file to where I need it. I suppose I could write a script that would do it.

**Comment by knz** [comment]
> there is a small inconvenience of having to run that script to create the user and having to supply the randomized username/password.

You can automate this by running the script with 

```
cockroach demo --listening-url-file=server.url
```

(assuming the code in #68706 )
then

```
cockroach sql --url `cat server.url` -f yourscript.sql
```

you can even make them auto-synchronize as follows:

```
# prepare a FIFO for synchronization
rm -f server.url; mkfifo server.url
# start demo in background
cockroach demo --listening-url-file=server.url -e 'select pg_sleep(10000000)' &
# start your init script automatically after the demo cluster has finished initializing
cockroach sql --url `cat server.url` -f yourscript.sql
```

**Comment by knz** [comment]
We're going to improve this by making secure mode easier to use, instead of making `--insecure` better.
The epic for this is CRDB-22814.

**Comment by knz** [comment]
actually the fix will also fix that.

**Comment by knz** [comment]
xref #94323
